cub/cub/detail/rfa.cuh

```diff
@@ -0,0 +1,950 @@
+
+// disable nan / infinity checks
+#  define DISABLE_NANINF
+#endif
+
+// jump table for indexing into data
+#define _CUB_RFA_MAX_JUMP_ 5
+static_assert(_CUB_RFA_MAX_JUMP_ <= 5, "MAX_JUMP greater than max");
+
+CUB_NAMESPACE_BEGIN
+
+namespace detail
+{
+namespace rfa
+{
+
+template <typename Float, int LEN>
+static _CCCL_DEVICE Float* get_shared_bin_array()
+{
+  static __shared__ Float bin_computed_array[LEN];
+  return bin_computed_array;
+}
+
+// vector types with RFA are not supported currently
+#if 0
+template <class T>
+struct get_vector_type
+{};
+template <>
+struct get_vector_type<double>
+{
+  using type = double2;
+};
+template <>
+struct get_vector_type<float>
+{
+  using type = float4;
+};
+
+template <class T>
+using vector_t = typename get_vector_type<T>::type;
+
+template <class T>
+auto vector_size()
+{
+  return sizeof(vector_t<T>) / sizeof(T);
+}
+
+template <class T>
+_CCCL_FORCEINLINE static double abs_max(const T&);
+
+template <>
+_CCCL_DEVICE double abs_max(const float4& x)
+{
+  return fmax(fmaxf(fabs(x.x), fabs(x.y)), fmax(fabs(x.z), fabs(x.w)));
+}
+
+template <>
+_CCCL_DEVICE double abs_max(const double4& x)
+{
+  return fmax(fmax(fabs(x.x), fabs(x.y)), fmax(fabs(x.z), fabs(x.w)));
+}
+
+template <>
+_CCCL_DEVICE double abs_max(const double2& x)
+{
+  return fmax(fabs(x.x), fabs(x.y));
+}
+
+template <>
+_CCCL_DEVICE double abs_max(const float2& x)
+{
+  return fmax(fabs(x.x), fabs(x.y));
+}
+#endif
+
+template <class ftype>
+struct RFA_bins
+{
+  static constexpr int BIN_WIDTH = ::cuda::std::is_same_v<ftype, double> ? 40 : 13;
+  static constexpr int MIN_EXP   = ::cuda::std::numeric_limits<ftype>::min_exponent;
+  static constexpr int MAX_EXP   = ::cuda::std::numeric_limits<ftype>::max_exponent;
+  static constexpr int MANT_DIG  = ::cuda::std::numeric_limits<ftype>::digits;
+
+  /// Binned floating-point maximum index
+  static constexpr int MAXINDEX = ((MAX_EXP - MIN_EXP + MANT_DIG - 1) / BIN_WIDTH) - 1;
+  // The maximum floating-point fold supported by the library
+  static constexpr int MAXFOLD = MAXINDEX + 1;
+
+  _CCCL_DEVICE static ftype initialize_bins(int index)
+  {
+    if (index == 0)
+    {
+      if constexpr (::cuda::std::is_same_v<ftype, float>)
+      {
+        return ::cuda::std::ldexp(0.75, MAX_EXP);
+      }
+      else
+      {
+        return 2.0 * ::cuda::std::ldexp(0.75, MAX_EXP - 1);
+      }
+    }
+
+    if (index > 0 && index <= MAXINDEX)
+    {
+      return ::cuda::std::ldexp(0.75, MAX_EXP + MANT_DIG - BIN_WIDTH + 1 - index * BIN_WIDTH);
+    }
+    else
+    {
+      return ::cuda::std::ldexp(0.75, MAX_EXP + MANT_DIG - BIN_WIDTH + 1 - MAXINDEX * BIN_WIDTH);
+    }
+  }
+};
+
+//! Class to hold a reproducible summation of the numbers passed to it
+//!
+//!@param ftype Floating-point data type; either `float` or `double
+//!@param FOLD  The fold, uses 3 by default.
+template <class ftype_,
+          int FOLD                                                                     = 3,
```

Comment:
Nitpick: `FOLD` should be spelled `Fold` but its not very descriptive of what it does